### ISSUE TITLE
Difference in output from decoding validation set while training and with marian-decoder

### ISSUE THREAD

**Issue by phikoehn** [marian-nmt/marian#154]
A new (to me, at least) way to run validation during decoding is the "translation" option for --valid-metrics. However, its output differs from running marian-decoder on the dumped model.

```
diff decoder.out model/model.iter220000.npz.bpe | more
3c3
< the two wanted to talk about the implementation of the international agreement and on T@@ ehr@@ an &apos;s dest@@ abili@@ zing measures in the Middle East .
---
> the two wanted to talk about the implementation of the international agreement and on T@@ ehr@@ an &apos;s dest@@ abili@@ zing actions in the Middle East .
6,10c6,10
```

Difference is "measures" instead of "actions".

`model/model.iter220000.npz.bpe` is a copy of the file specified with "--valid-translation-output" during training.

`decoder.out2` is the output of

```
$marian/build/marian-decoder \
        --model model/model.npz \
        --devices $GPU \
        --vocabs data/train.bpe.de.json data/train.bpe.en.json \
        -i data/dev.bpe.de \
        --log model/evaluate-final-model.log \
        > decoder.out
```

**Comment by emjotde** [comment]
Hi,
since both of them are using exactly the same functions I would guess there might be some options mismatch for translation options for marian and marian-decoder.

Can you post your full training config? Also can you try 
``
marian-decoder -c model/model.npz.decoder.yml 
``
instead of specifying just `-m model.npz` to check for differences?

**Comment by phikoehn** [comment]
Adding `-c model/model.npz.decoder.yml` makes no difference.

Training:

```
$marian/build/marian \
        --model model/model.npz \
        --devices $GPU \
        --train-sets data/train.bpe.de data/train.bpe.en \
        --vocabs data/train.bpe.de.json data/train.bpe.en.json \
        --mini-batch-fit -w 3000 \
        --dim-vocabs 50000 50000 \
        --layer-normalization --dropout-rnn 0.2 --dropout-src 0.1 --dropout-trg 0.1 \
        --learn-rate 0.0001 \
        --after-epochs 0 \
        --early-stopping 5 \
        --valid-freq 20000 --save-freq 20000 --disp-freq 2000 \
        --valid-mini-batch 8 \
        --valid-sets data/dev.bpe.de data/dev.bpe.en \
        --valid-metrics cross-entropy perplexity translation \
        --valid-translation-output model/dev.out \
        --valid-script-path $validate \
        --seed 1111 --exponential-smoothing \
        --normalize=1 --beam-size=12 --quiet-translation \
        --log model/train.log --valid-log model/valid.log
```

Decoding by hand:

```
$marian/build/marian-decoder \
        --model model/model.npz \
        --devices $GPU \
        --vocabs data/train.bpe.de.json data/train.bpe.en.json \
        -i data/dev.bpe.de \
        --log model/evaluate-final-model.log \
        > model/dev.out
```

**Comment by emjotde** [comment]
How about adding `--normalize=1 --beam-size=12 ` to the decoder?

**Comment by phikoehn** [comment]
This gives different output again (which it really should not), but it is also different from the training validation output.

**Comment by emjotde** [comment]
This should be more similar to the validation output. Small differences could be expected due to floating point inconsistencies with batched decoding. How about adding `--mini-batch 8` to the decoder?

**Comment by phikoehn** [comment]
`--mini-batch 8` makes no difference.

Reported config when running decoder:

```
[2018-01-18 11:12:34] [config] allow-unk: false
[2018-01-18 11:12:34] [config] beam-size: 12
[2018-01-18 11:12:34] [config] best-deep: false
[2018-01-18 11:12:34] [config] dec-cell: gru
[2018-01-18 11:12:34] [config] dec-cell-base-depth: 2
[2018-01-18 11:12:34] [config] dec-cell-high-depth: 1
[2018-01-18 11:12:34] [config] dec-depth: 1
[2018-01-18 11:12:34] [config] devices:
[2018-01-18 11:12:34] [config]   - 2
[2018-01-18 11:12:34] [config] dim-emb: 512
[2018-01-18 11:12:34] [config] dim-rnn: 1024
[2018-01-18 11:12:34] [config] dim-vocabs:
[2018-01-18 11:12:34] [config]   - 50000
[2018-01-18 11:12:34] [config]   - 50000
[2018-01-18 11:12:34] [config] enc-cell: gru
[2018-01-18 11:12:34] [config] enc-cell-depth: 1
[2018-01-18 11:12:34] [config] enc-depth: 1
[2018-01-18 11:12:34] [config] enc-type: bidirectional
[2018-01-18 11:12:34] [config] ignore-model-config: false
[2018-01-18 11:12:34] [config] input:
[2018-01-18 11:12:34] [config]   - data/dev.bpe.de
[2018-01-18 11:12:34] [config] layer-normalization: true
[2018-01-18 11:12:34] [config] log: decoder.log
[2018-01-18 11:12:34] [config] log-level: info
[2018-01-18 11:12:34] [config] max-length: 1000
[2018-01-18 11:12:34] [config] max-length-crop: false
[2018-01-18 11:12:34] [config] maxi-batch: 1
[2018-01-18 11:12:34] [config] maxi-batch-sort: none
[2018-01-18 11:12:34] [config] mini-batch: 1
[2018-01-18 11:12:34] [config] models:
[2018-01-18 11:12:34] [config]   - model/model.iter220000.npz
[2018-01-18 11:12:34] [config] n-best: false
[2018-01-18 11:12:34] [config] normalize: 0
[2018-01-18 11:12:34] [config] port: 8080
[2018-01-18 11:12:34] [config] quiet: false
[2018-01-18 11:12:34] [config] quiet-translation: false
[2018-01-18 11:12:34] [config] relative-paths: false
[2018-01-18 11:12:34] [config] right-left: false
[2018-01-18 11:12:34] [config] seed: 0
[2018-01-18 11:12:34] [config] skip: false
[2018-01-18 11:12:34] [config] tied-embeddings: false
[2018-01-18 11:12:34] [config] tied-embeddings-all: false
[2018-01-18 11:12:34] [config] tied-embeddings-src: false
[2018-01-18 11:12:34] [config] transformer-dim-ffn: 2048
[2018-01-18 11:12:34] [config] transformer-heads: 8
[2018-01-18 11:12:34] [config] transformer-postprocess: dan
[2018-01-18 11:12:34] [config] transformer-postprocess-emb: d
[2018-01-18 11:12:34] [config] transformer-preprocess: ""
[2018-01-18 11:12:34] [config] type: amun
[2018-01-18 11:12:34] [config] version: v1.0.0+38d4bd0
[2018-01-18 11:12:34] [config] vocabs:
[2018-01-18 11:12:34] [config]   - data/train.bpe.de.json
[2018-01-18 11:12:34] [config]   - data/train.bpe.en.json
[2018-01-18 11:12:34] [config] workspace: 512
[2018-01-18 11:12:34] [config] Model created with Marian v1.0.0+38d4bd0
```

Reported config when running training

```
[2018-01-17 18:31:52] [config] after-batches: 0
[2018-01-17 18:31:52] [config] after-epochs: 0
[2018-01-17 18:31:52] [config] allow-unk: false
[2018-01-17 18:31:52] [config] batch-flexible-lr: false
[2018-01-17 18:31:52] [config] batch-normal-words: 1920
[2018-01-17 18:31:52] [config] beam-size: 12
[2018-01-17 18:31:52] [config] best-deep: false
[2018-01-17 18:31:52] [config] clip-norm: 1
[2018-01-17 18:31:52] [config] cost-type: ce-mean
[2018-01-17 18:31:52] [config] dec-cell: gru
[2018-01-17 18:31:52] [config] dec-cell-base-depth: 2
[2018-01-17 18:31:52] [config] dec-cell-high-depth: 1
[2018-01-17 18:31:52] [config] dec-depth: 1
[2018-01-17 18:31:52] [config] devices:
[2018-01-17 18:31:52] [config]   - 0
[2018-01-17 18:31:52] [config] dim-emb: 512
[2018-01-17 18:31:52] [config] dim-rnn: 1024
[2018-01-17 18:31:52] [config] dim-vocabs:
[2018-01-17 18:31:52] [config]   - 50000
[2018-01-17 18:31:52] [config]   - 50000
[2018-01-17 18:31:52] [config] disp-freq: 2000
[2018-01-17 18:31:52] [config] dropout-rnn: 0.2
[2018-01-17 18:31:52] [config] dropout-src: 0.1
[2018-01-17 18:31:52] [config] dropout-trg: 0.1
[2018-01-17 18:31:52] [config] early-stopping: 5
[2018-01-17 18:31:52] [config] embedding-fix-src: false
[2018-01-17 18:31:52] [config] embedding-fix-trg: false
[2018-01-17 18:31:52] [config] embedding-normalization: false
[2018-01-17 18:31:52] [config] enc-cell: gru
[2018-01-17 18:31:52] [config] enc-cell-depth: 1
[2018-01-17 18:31:52] [config] enc-depth: 1
[2018-01-17 18:31:52] [config] enc-type: bidirectional
[2018-01-17 18:31:52] [config] exponential-smoothing: 0.0001
[2018-01-17 18:31:52] [config] grad-dropping-momentum: 0
[2018-01-17 18:31:52] [config] grad-dropping-rate: 0
[2018-01-17 18:31:52] [config] grad-dropping-warmup: 100
[2018-01-17 18:31:52] [config] guided-alignment-cost: ce
[2018-01-17 18:31:52] [config] guided-alignment-weight: 1
[2018-01-17 18:31:52] [config] ignore-model-config: false
[2018-01-17 18:31:52] [config] keep-best: false
[2018-01-17 18:31:52] [config] label-smoothing: 0
[2018-01-17 18:31:52] [config] layer-normalization: true
[2018-01-17 18:31:52] [config] learn-rate: 0.0001
[2018-01-17 18:31:52] [config] log: model/train.log
[2018-01-17 18:31:52] [config] log-level: info
[2018-01-17 18:31:52] [config] lr-decay: 0
[2018-01-17 18:31:52] [config] lr-decay-freq: 50000
[2018-01-17 18:31:52] [config] lr-decay-inv-sqrt: 0
[2018-01-17 18:31:52] [config] lr-decay-repeat-warmup: false
[2018-01-17 18:31:52] [config] lr-decay-reset-optimizer: false
[2018-01-17 18:31:52] [config] lr-decay-start:
[2018-01-17 18:31:52] [config]   - 10
[2018-01-17 18:31:52] [config]   - 1
[2018-01-17 18:31:52] [config] lr-decay-strategy: epoch+stalled
[2018-01-17 18:31:52] [config] lr-report: false
[2018-01-17 18:31:52] [config] lr-warmup: 0
[2018-01-17 18:31:52] [config] lr-warmup-at-reload: false
[2018-01-17 18:31:52] [config] lr-warmup-cycle: false
[2018-01-17 18:31:52] [config] lr-warmup-start-rate: 0
[2018-01-17 18:31:52] [config] max-length: 50
[2018-01-17 18:31:52] [config] max-length-crop: false
[2018-01-17 18:31:52] [config] maxi-batch: 100
[2018-01-17 18:31:52] [config] maxi-batch-sort: trg
[2018-01-17 18:31:52] [config] mini-batch: 64
[2018-01-17 18:31:52] [config] mini-batch-fit: true
[2018-01-17 18:31:52] [config] mini-batch-words: 0
[2018-01-17 18:31:52] [config] model: model/model.npz
[2018-01-17 18:31:52] [config] n-best: false
[2018-01-17 18:31:52] [config] no-reload: false
[2018-01-17 18:31:52] [config] no-shuffle: false
[2018-01-17 18:31:52] [config] normalize: 1
[2018-01-17 18:31:52] [config] optimizer: adam
[2018-01-17 18:31:52] [config] optimizer-delay: 1
[2018-01-17 18:31:52] [config] overwrite: false
[2018-01-17 18:31:52] [config] quiet: false
[2018-01-17 18:31:52] [config] quiet-translation: true
[2018-01-17 18:31:52] [config] relative-paths: false
[2018-01-17 18:31:52] [config] right-left: false
[2018-01-17 18:31:52] [config] save-freq: 20000
[2018-01-17 18:31:52] [config] seed: 1111
[2018-01-17 18:31:52] [config] skip: false
[2018-01-17 18:31:52] [config] sync-sgd: false
[2018-01-17 18:31:52] [config] tempdir: /tmp
[2018-01-17 18:31:52] [config] tied-embeddings: false
[2018-01-17 18:31:52] [config] tied-embeddings-all: false
[2018-01-17 18:31:52] [config] tied-embeddings-src: false
[2018-01-17 18:31:52] [config] train-sets:
[2018-01-17 18:31:52] [config]   - data/train.bpe.de
[2018-01-17 18:31:52] [config]   - data/train.bpe.en
[2018-01-17 18:31:52] [config] transformer-dim-ffn: 2048
[2018-01-17 18:31:52] [config] transformer-dropout: 0
[2018-01-17 18:31:52] [config] transformer-dropout-attention: 0
[2018-01-17 18:31:52] [config] transformer-heads: 8
[2018-01-17 18:31:52] [config] transformer-postprocess: dan
[2018-01-17 18:31:52] [config] transformer-postprocess-emb: d
[2018-01-17 18:31:52] [config] transformer-preprocess: ""
[2018-01-17 18:31:52] [config] type: amun
[2018-01-17 18:31:52] [config] valid-freq: 20000
[2018-01-17 18:31:52] [config] valid-log: model/valid.log
[2018-01-17 18:31:52] [config] valid-max-length: 1000
[2018-01-17 18:31:52] [config] valid-metrics:
[2018-01-17 18:31:52] [config]   - cross-entropy
[2018-01-17 18:31:52] [config]   - perplexity
[2018-01-17 18:31:52] [config]   - translation
[2018-01-17 18:31:52] [config] valid-mini-batch: 8
[2018-01-17 18:31:52] [config] valid-script-path: ./score-dev.sh
[2018-01-17 18:31:52] [config] valid-sets:
[2018-01-17 18:31:52] [config]   - data/dev.bpe.de
[2018-01-17 18:31:52] [config]   - data/dev.bpe.en
[2018-01-17 18:31:52] [config] valid-translation-output: model/dev.out
[2018-01-17 18:31:52] [config] vocabs:
[2018-01-17 18:31:52] [config]   - data/train.bpe.de.json
[2018-01-17 18:31:52] [config]   - data/train.bpe.en.json
[2018-01-17 18:31:52] [config] workspace: 3000
```

**Comment by emjotde** [comment]
And I just realized that the saved model may not be the same model as the one that is doing validation. This is async SGD. So between the validation step and the saving it could finish another training batch.

**Comment by phikoehn** [comment]
Okay. Not sure if it matters much. But if then, that would be a problem, since the final model is selected based on the reported BLEU scores.

**Comment by emjotde** [comment]
What differences are we talking here? 3 words in 3000 sentences is expected. If more, something is wrong.

**Comment by phikoehn** [comment]
BLEU score is 33.44 (marian-decoder) vs. 33.97 (training validation). I guess that does matter.

**Comment by phikoehn** [comment]
Almost all sentences are different.

**Comment by emjotde** [comment]
Yes. That must be something different. Can you post the content of model.npz.decoder.yml ?

**Comment by emjotde** [comment]
For the decoder:
``
[2018-01-18 11:12:34] [config] normalize: 0
``

try again `--normalize=1`

**Comment by phikoehn** [comment]
The BLEU scores I gave you are with normalize 1

**Comment by emjotde** [comment]
Hm. Will have to try to reproduce that. Is this WMT de-en? Usually I get the same scores.

**Comment by emjotde** [comment]
I get identical scores for all tested models, different language pairs and directions, all like reported in valid.log. 

Let's check again. Can you post the content of `model/valid.log`?
And can you run:

```
./marian-decoder -c model/marian.npz.decoder.yml -m model/model.iter220000.npz -i data/dev.bpe.de > dev.out
./score-dev.sh dev.out
```

**Comment by phikoehn** [comment]
Hmmm... I guess that is good news and bad news. 

```
[...]
[2018-01-18 07:42:12] [valid] 180000 : translation : 33.73 : new best
[2018-01-18 09:09:02] [valid] 200000 : cross-entropy : 41.9101 : new best
[2018-01-18 09:09:10] [valid] 200000 : perplexity : 5.21429 : new best
[2018-01-18 09:09:50] [valid] 200000 : translation : 33.75 : new best
[2018-01-18 10:36:36] [valid] 220000 : cross-entropy : 41.5066 : new best
[2018-01-18 10:36:45] [valid] 220000 : perplexity : 5.13205 : new best
[2018-01-18 10:37:25] [valid] 220000 : translation : 33.97 : new best
[...]
```
versus

```
/opt/nmt/marian-v1.1/build/marian-decoder -c model/model.npz.decoder.yml -m model/model.iter220000.npz -i data/dev.bpe.de --devices 2 > dev.out
cat dev.out | sed 's/\@\@ //g' | /opt/nmt/mosesdecoder/scripts/generic/multi-bleu.perl data/dev.tok.e
BLEU = 33.44, 67.2/41.1/27.0/18.1 (BP=0.980, ratio=0.981, hyp_len=64370, ref_len=65647)
```

**Comment by emjotde** [comment]
Hm, maybe this is indeed an issue with asynchronous SGD which I have not been using for a while (I mostly use --sync-sgd now). Can you make training data/configs/vocabs available? Maybe also the failing models.

**Comment by phikoehn** [comment]
Sure - this is straight-up WMT German-English.

I up putting it up at http://www.statmt.org/shallow.tgz - it will take ~20 minutes for me to upload.

**Comment by phikoehn** [comment]
& it's up

**Comment by emjotde** [comment]
So, if I look at the code, it seems indeed that there are updates happening between saving and validation in asynchronous multi-gpu mode. This is new as for type translation there is now a barrier that requires all gpus to finish before validation. Need to think a moment how to fix that.

**Comment by phikoehn** [comment]
By the way, the reason I looked into that, is because I had to run validation by hand, when training finished due to "max-epoch" setting, not validation convergence. There must be an easy hack to trigger validation on the final model no matter what.

**Comment by emjotde** [comment]
Oh, good idea. @snukky can you take a look at this? This needs to be done in all graph groups separately.

**Comment by emjotde** [comment]
@snukky, actually it can be done in training.h for all groups. Moving average may make it more complicated. Maybe we need to expose something like `model->validate()`.

**Comment by emjotde** [comment]
@phikoehn This should be fixed in [marian-dev](https://github.com/marian-nmt/marian-dev) now. Will update soon and close here.

**Comment by snukky** [comment]
Based on the issue, what scenarios we should transform into regression tests?

**Comment by emjotde** [comment]
@snukky I would make sure that the model that gets saved per iteration based on --save-freq is the same as the model saved with --keep-best (model.npz.best-translation) after a couple of iterations in async multi-gpu mode. We do not need to check the translation, just if the models are equal. On the other hand, checking translation does not hurt, either. 

For the evaluating the final model thing I will open an issue in marian-dev.
